### PR TITLE
Feat/release ingest

### LINKED ISSUE #331: Ingest releases per repo: cadence chart and staleness columns
> 🧑‍💻 **Intermediate Issue** — a complex task spanning multiple modules, with real design decisions to own.
> **Time:** ~25 hours · **Prerequisites:** comfortable navigating this repo (a completed [beginner issue](https://github.com/hiero-hackers/analytics/issues?q=is%3Aissue+state%3Aopen+no%3Aassignee+label%3Abeginner) is the usual route; demonstrated CI/CD proficiency substitutes for workflow-focused issues).
> We expect more than "it works": maintainable code that fits the existing architecture.

## The task

**Problem:** The dashboard ingests PRs, issues, contributors, roles, scorecards — but nothing about **releases**. It cannot answer "which repos ship regularly, which haven't cut a release in a year, and did the SDKs ship together after the last consensus-node release?" — questions both maintainers and the TSC actually ask. Shipping cadence is a health signal on par with contributor activity, and right now it's a blind spot.

**What done looks like:** A new ingest module fetches each org repo's releases (tag, name, published date, prerelease flag), a new registered pipeline turns them into:
- a per-repo releases table (CSV artifact),
- a cadence chart on the dashboard,
- and staleness columns joined into the existing repo-summaries view — at minimum `latest_release` and `days_since_last_release`.

Repos with **zero** releases must appear as maximally stale rather than disappearing — the honest-denominator principle `hip_repo_engagement` already follows. Offline runs skip cleanly, like every other network pipeline.

**Visualisation** — release data is sparse event data, so avoid per-repo line charts (mostly zeros) in favour of:
- Headline: a **dot/strip timeline** — repos on the y-axis, release dates on the x-axis over the last ~18 months, prereleases visually distinct. Cadence, droughts, and coordinated release trains all show up as visual texture; it's a plain matplotlib scatter using the house style in `plotting/base.py`/`plotting/style.py`, no new primitive family needed.
- Secondary (optional): a **days-since-last-release bar** per repo, which doubles as the visual form of the staleness column.

**Modules involved / constraints:**
- `data_sources/github_ingest/` — new module, following the conventions of the existing ones (`issues.py`, `pull_requests.py`); the GraphQL query belongs with its siblings in `data_sources/queries.py`. Decide whether incremental fetching (`incremental.py` pattern) is worth it — release lists are small, so a full refetch per run may be the simpler right answer; say which you chose and why.
- Scope decision to own: many repos cut **git tags without GitHub Releases**. Recommendation: GitHub Releases only for this issue, with the tag-fallback documented as an explicit non-goal / follow-up — mixing the two data sources muddies "release" semantics and can double-count. Drafts excluded; prereleases included but flagged.
- New pipeline registered in `pipelines/` (registry entry gives you the CLI subcommand for free); staleness columns land via `analysis/repo_summaries.py`; the new chart and any new section are declared in `dashboard_spec/`, which is what makes them appear on the dashboard.
- The output-contract test pins the output surface — new CSVs/charts are a deliberate contract update. Chart colours belong in `config/charts.py`, not inline.

## How to work on this

1. **Claim it:** comment `/assign` and wait to be assigned — unassigned PRs are closed automatically.
2. **Get a plan:** once assigned, comment `@coderabbitai plan` for a draft plan, then do your own investigation — [docs/architecture.md](https://github.com/hiero-hackers/analytics/blob/main/docs/architecture.md) maps the layers and their rules.
3. **Propose your approach as a comment before coding.** A paragraph is enough; early feedback here routinely saves days of rework.

**🤖 AI:** tools are welcome; verified work is required — you can explain every line and defend every design choice. See the [AI policy](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#ai-policy). Fully automated bot PRs are closed.

**Worth knowing about this repo before you design:**

- The layer rules in [docs/architecture.md](https://github.com/hiero-hackers/analytics/blob/main/docs/architecture.md) are strict — review will hold your solution to them.
- Tests mirror src (`tests/<pkg>/test_<module>.py`), and the output-contract test pins the pipeline output surface — if your change adds or renames outputs, update the contract deliberately.

**Before opening your PR:**

- [ ] I proposed my approach on this issue and incorporated any feedback
- [ ] The solution fits the existing architecture and layer rules, and is clear enough for others to debug without me
- [ ] Tests cover the happy path, edge cases, and error handling ([testing guide](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#testing))
- [ ] I reviewed my own diff line by line; scope is limited to this issue
- [ ] Workflow checks pass — CI green, signed commits, linked issue

**Stuck?** Comment here with what you've tried — see [getting help](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#getting-help).

### PR DESCRIPTION
**Pull request by iron-prog**
**Description**:

Add GitHub Releases analytics to track release cadence and repository staleness.

* Add GitHub Releases GraphQL ingestion with caching and prerelease support.
* Add release timeline and per-repository staleness analysis, including cadence-relative `staleness_ratio`.
* Add release timeline and staleness visualizations.
* Add a dedicated Releases dashboard with release-health KPIs and repository summary data.
* Add tests for ingestion, analysis, pipeline, plotting, dashboard contracts, and KPI metrics.

**Related issue(s)**:

Fixes #331

**Notes for reviewer**:

The implementation keeps release data independent from the governance/period-scoped repository overview and preserves repositories with no releases in the summary.

`staleness_ratio` compares the current release gap with each repository's historical median release gap, making staleness relative to its own cadence rather than raw days alone.

Validated against real `hiero-ledger` data. The pipeline successfully generated the release timeline/summary CSVs and both release charts.

* `752 passed`
* `95.40%` coverage

**Checklist**

* [x] I claimed the linked issue with `/assign` before starting (see [[contributing guide](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#workflow)](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#workflow))
* [x] `uv run pytest` and `uv run ruff check src tests` pass locally
* [x] Tests added/updated for the change (mirroring the `src/` layout)
* [x] Commits are signed and signed-off: `git commit -S -s`
* [x] Docs updated if relevant

### CODE REVIEW

**Comment by iron-prog** [comment]
Status update: both CodeRabbit findings are addressed.

- **Pagination bound:** fixed  and resolved.
- **Never-released repos:** `staleness_ratio` is now `inf`, while repos with only one release remain `null`. Tables/CSVs rank never-released repos first; the staleness chart excludes infinite values and documents this explicitly.

One note: the Linked Issues check flags the missing `repo_activity_overview` join. This is expected because the design was later redirected away from that approach during #331.

Ready for another review.

**Comment by iron-prog** [comment]
@exploreriii any update on this

**Comment by exploreriii** [comment]
Hi this is a complicated PR to review, please feel free to take on another issue while you wait :)
Thanks for understanding

**Review by exploreriii — COMMENTED** [review]
Hi @iron-prog 
I really like your approach

Just a couple extra requirements

Image

Could you please check using F12 on different screen sizes to see the chart renders as viewable as possible? ideally it would stretch to the viewport so it is not small 
Further have you considered cahnging the x axis so it is more human friendly e.g. Nov with 2025 underneath or something? 

Ideally we can make the release timeline dynamic - we have time filters in other charts and tables (1 month, 1 year, etc). can we use them so the reader can change what they want to see?
Please also check in dark mode, the colours render with enough contrast

I personally find the second bar chart overwhelming - Shipping repos most overdue relative to their own pace
I actually think it has similar data to the first that the user can 'absorb' visually anyway
So, does the second chart really present a different view, and if it does, in an easy to undesrtand way? I think mathematically it is nice, but for the use case it might not work well

table 3 covers that well too. maybe instead we can delete chart 2, keep table 3 and perhaps make the  overdue vs. own pace column more visual? like for example something that is doing late releases can be a certain colour different to the avearge 

Another idea is to consider groupign the repos by type - simialr to the network graphs : sdks, core, etc. That might help us gain some insight, but it might be overkill

What are your thoughts? i think its pretty good already! nice work

**Comment by exploreriii** [comment]
Also please do check hiero hackers will load correctly, and also that each of the new content has a widget url, as introduced by @MonaaEid recently

**Comment by iron-prog** [comment]
Thanks for the feedback! I agree with the suggestions. I’ll make the timeline responsive with better date formatting and dark-mode support, and I’ll also address  chart problem and make the overdue-vs-own-pace column more visual in the table.

**Review by MonaaEid — COMMENTED** [review]
Great work! one question why isn’t there releases analytics for Hiero hackers?

**Comment by iron-prog** [comment]
> Great work! one question why isn’t there releases analytics for Hiero hackers?

Thanks! This should be fixed now — the latest changes render Releases analytics for Hiero Hackers as well.

**Comment by MonaaEid** [comment]
@exploreriii should we include release analytics for Hiero-hackers too?

**Comment by exploreriii** [comment]
sure, there will be a lot of nulls, which will hav eto think about how to present

File: src/hiero_analytics/pipelines/releases.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+def main(org: str = ORG):
+    """Fetch releases for every repo in the org and publish the timeline/staleness tables + charts."""
+    client, org_data_dir, org_charts_dir = org_context(org)
+
+    all_repos = fetch_org_repos_graphql(client, org)
+    if not all_repos:
+        logger.warning("No repositories found for org: %s", org)
+        return
+
+    records = fetch_org_releases_graphql(client, org)
```
Comment (exploreriii) [inline src/hiero_analytics/pipelines/releases.py]: i think you are fetching the org repo list twice, possible this causes some problems

File: src/hiero_analytics/export/macro_metrics.py [hunk 2]
```diff
@@ -105,7 +105,43 @@ def governance_metrics(loaded: dict[str, pd.DataFrame], org_data_dir: Path) -> l
...
+    if total_releasing == 0:
+        return [("repos with releases", f"0 of {len(summary)}")]
+
+    metrics: list = [("repos with releases", f"{total_releasing} of {len(summary)}")]
+
+    recent = int((pd.to_numeric(releasing["days_since_last_release"], errors="coerce") <= 90).sum())
+    metrics.append(("released last 90d %", _pct(recent, total_releasing)))
+
+    ratio = pd.to_numeric(releasing["staleness_ratio"], errors="coerce")
+    overdue = int((ratio > 3).sum())
```
Comment (exploreriii) [inline src/hiero_analytics/export/macro_metrics.py]: shouldn't be hard coding things please when you have the variables already defined

File: src/hiero_analytics/data_sources/github_ingest/releases.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+) -> list[ReleaseRecord]:
+    """Fetch every published, non-draft release for one repository.
+
+    GitHub orders the connection by ``CREATED_AT`` descending, but pagination
+    always runs to completion — releases are cheap enough per repo that there
+    is no early-stop cutoff, unlike the high-volume PR/issue fetchers.
+    """
+    cache_scope = f"{owner}_{repo}"
+    cache_parameters = {"owner": owner, "repo": repo}
+    cached = load_records_cache(
```
Comment (exploreriii) [inline src/hiero_analytics/data_sources/github_ingest/releases.py:95]: you are hand rolling cache sequencing that the shared fetch github resource i think already does, please check

File: src/hiero_analytics/data_sources/github_ingest/releases.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    def fetch_func(repo):
+        """Fetch releases for a single repository."""
+        return fetch_repo_releases_graphql(
+            client,
+            repo.owner,
+            repo.name,
+            **_cache_kwargs(use_cache, cache_ttl_seconds, refresh),
+        )
+
+    return fetch_org_resource_parallel(
```
Comment (exploreriii) [inline src/hiero_analytics/data_sources/github_ingest/releases.py]: are you sure you need to use fetch_org_resource_parallel rather than fetch_org_records_batched, consider API limits

**Review by exploreriii — DISMISSED** [review]
I think there are some edge cases here to refine over time once we get more solid ideas on how this would play out. For instance, repos that never will do a release (not applicable to them), repos that release multiple times in a day, pre releases counting as releases.

I would definitely like to see equal deevelopment approach at hiero hackers, ideally the dashboad doesn't differentiate by org, it presents the same data to each (we havn't done this yet but should do, just in case we add other orgs one day)

**Comment by exploreriii** [comment]
@iron-prog if you can help us with completing reviews and creating relevant issues, we are looking for new committers to the analytics project that would require a mix of intermediate skills (triaging issues, reviewing prs, creating prs)

**Comment by exploreriii** [comment]
Some notes on visuals - ideally we keep a similar tabbing option to the rest of the dashboard.
These explanations need to be user focused, the rest can go in a methodology note. Every release per repo, windowed by the tabs below (Week / 1 month / 1 year / Last 18 months) or consider a scroll format. Y-axis sorted by release count in the selected window (busiest at top); the number beside each row is that count, so a high-cadence repo's row stays readable instead of turning into a solid smear of overlapping dots.
Each chart should open up and show an expandable methodoloy, maybe we destroyed this feature recently by accident.

Release staleness by repo may have to separate out pre releases maybe in another table, or at least add some notes.
Similarly, this is not user centric Every repo in the org, whether or not it has ever released — a repo with no releases still gets a row rather than being silently absent, and its 'pace' column ranks it as maximally stale. 'staleness_ratio' is blank for the narrower case of a repo with only one release ever, or a same-day repeat release — it has shipped, there just isn't enough history yet to say what's normal for it.

We can definitely round e.g. overdue numbers for clarity

Pace isn't sorting correctly, i.e. never released isn't sorting to the bottom or top when ranked, this is to do with bugs in the null handling throughout the analytics pipeline. never released and overdue are the same colours, probably incorrreclty so.

Please check the visuals! the repo 'standards' should always be improved - we have gone from proof of concept to someething more solid, so really if there is an opportunity to improve, we should

**Comment by iron-prog** [comment]
Thanks for the detailed feedback. I’ll work through the points and adjust the PR accordingly. my university exams are currently going on, so my availability is a bit limited this week. I’ll be more available after this week and would be happy to contribute more actively, including reviews, issue triage and working toward a committer role in the project.

**Review by danielmarv — COMMENTED** [review]
your link checks are failing

**Comment by iron-prog** [comment]
> your link checks are failing

I m going to fix it

File: web/src/app.css [hunk 5]
```diff
@@ -838,6 +838,14 @@
...
     color: var(--soft);
   }
+  .chip-overdue {
+    --dot: var(--neg);
+    color: var(--neg);
+  }
+  .chip-watch {
+    --dot: var(--warn);
+    color: var(--warn-ink);
+  }
```
Comment (MonaaEid) [inline web/src/app.css:848]: the staleness renderer maps buckets to CSS classes:
```
 never_released: 'chip-never',
overdue: 'chip-overdue',
watch: 'chip-watch',
on_pace: 'chip-merged',
insufficient_history: 'chip-none',
```
I only see `.chip-overdue` and `.chip-watch` added in app.css... if you prefer reusing existing classes, update the mapping in FormattedCell.tsx to use the project's canonical chip class names

File: src/hiero_analytics/data_sources/github_ingest/releases.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        return cached_records
+
+    batched_records, failed_repos = fetch_repos_batched(
+        client,
+        stale_repos,
+        query_text=load_query("releases"),
+        model_class=ReleaseRecord,
+        nodes_path=["releases"],
+        context_builder=lambda repo: {"owner": repo.owner, "repo": repo.name},
+        max_workers=max_workers,
```
Comment (MonaaEid) [inline src/hiero_analytics/data_sources/github_ingest/releases.py]:
```suggestion
```
this argument causes a TypeError which prevents the script from running properly

**Review by MonaaEid — DISMISSED** [review]
Hi @iron-prog, nice work.. just few comments:
the chart is meant to display "the busiest repos" at the top, but it’s currently showing the smallest-count repos first... can we fix that?
Image

**Comment by iron-prog** [comment]
Screenshot 2026-09-10 at 4 12 00 PM

Hi @MonaEid, sorry for the delayed response.
I was tied up with a project for the past few days.
Thanks for the review. I've addressed the feedback:

- The release timeline now correctly shows the busiest repositories at the top, sorted by release count.
- The staleness chip mapping now uses the existing canonical CSS classes `(chip-none,` `chip-overdue,` `chip-watch,` and `chip-merged).`
- I also addressed the retry behavior for repositories that fail during batched release 
Thanks again for the helpful feedback!

**Comment by iron-prog** [comment]
Hi @exploreriii  I’ve addressed the review feedback and fixed theissues.
When you get a chance, could you please take another look at the PR? Thanks!

**Review by danielmarv — APPROVED** [review]
LGTM!

**Comment by exploreriii** [comment]
thank you all for your help
